### PR TITLE
Add a _BinaryUfuncInfo class and use it to test some binary ufuncs

### LINKED ISSUE #54294: Port the basic binary operations (add, sub, mul and div, true_divide) to use OpInfos
This issue is a more specific part of the overall OpInfo porting issue https://github.com/pytorch/pytorch/issues/54261.

This will require resolving why we currently have both OpInfos and method_tests() entries for div. We should ensure all historic cases covered by method_tests() are covered by the OpInfos' sample inputs, too.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by mattip**
Fixes #54294

Use OpInfo classes to simplify testing of binary ufuncs. This PR starts with add, sub, mul and divide.

### CODE REVIEW

**Comment by mattip** [comment]
To make it easier to review, I put the new test in a separate file, the first commits only test "normal" values.

File: test/run_test.py [hunk 1]
```diff
@@ -34,7 +34,6 @@
     'test_type_hints',
     'test_autograd',
     'benchmark_utils/test_benchmark_utils',
-    'test_binary_ufuncs',
```
Comment (mruberry) [inline test/run_test.py:37]: Won't this cause test_binary_ufuncs to not run?

File: test/run_test.py [hunk 2]
```diff
@@ -256,6 +255,7 @@
     'test_jit_profiling',
     'test_torch',
     'test_binary_ufuncs'
+    'test_binary_ufuncs_ops'
```
Comment (mruberry) [inline test/run_test.py:258]: It's probably simpler to not bother with a new file

File: test/test_binary_ufuncs_ops.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                    continue
+
+            self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+    # Tests that the function and its (array-accepting) reference produce the same
+    #   values on a range of tensors, including empty tensors, scalar tensors,
+    #   1D tensors and a large 2D tensor with interesting and extremal values
+    #   and discontiguities.
+    @ops(binary_ufuncs)
+    def test_reference_numerics_normal(self, device, dtype, op):
```
Comment (mruberry) [inline test/test_binary_ufuncs_ops.py:224]: What if we start with just "test_reference_numerics_normal" and it simply calls make_tensor() to construct two large tensors of the same size for the moment?

Then we can think about how to extend the tests. I have some ideas that might simplify things.

This would let us remove the _hard and _extremal variants (for now) and cut the more complicated generator functions.

File: test/test_unary_ufuncs.py [hunk 4]
```diff
@@ -336,7 +336,6 @@ def _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan=Tru
     #   values on a range of tensors, including empty tensors, scalar tensors,
     #   1D tensors and a large 2D tensor with interesting and extremal values
     #   and discontiguities.
-    @suppress_warnings
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:339]: Why this change?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -306,7 +306,7 @@ class UnaryUfuncInfo(OpInfo):
       - they typically support the out kwarg
       - they typically have NumPy or SciPy references
     See NumPy's universal function documentation
-    (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+    (https://numpy.org/doc/stable/reference/ufuncs.html) for more details
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:309]: Thank you for fixing this.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -345,6 +345,58 @@ def __init__(self,
...
+    def __init__(self,
+                 name,  # the string name of the function
+                 *,
+                 ref,  # a reference function
+                 dtypes=floating_types(),
+                 dtypesIfCPU=floating_and_complex_types_and(torch.bfloat16),
+                 dtypesIfCUDA=floating_and_complex_types_and(torch.half),
+                 dtypesIfROCM=floating_types_and(torch.half),
+                 domain=(None, None),  # the [low, high) domain of the function
+                 handles_large_floats=True,  # whether the op correctly handles large float values (like 1e20)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:373]: We can probably get away without adding metadata like `handles_large_floats`, `handles_extremals`, `handles_complex_extremals` and `supports_complex_to_float`. 

(In fact we should probably remove all these from the UnaryUfuncInfos, too.)

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -345,6 +345,58 @@ def __init__(self,
...
+                 handles_extremals=True,  # whether the op correctly handles extremal values (like inf)
+                 handles_complex_extremals=True,  # whether the op correct handles complex extremals (like inf -infj)
+                 supports_complex_to_float=False,  # op supports casting from complex input to real output safely eg. angle
+                 sample_inputs_func=sample_inputs_unary,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:377]: Let's create sample_inputs_binary so the sample inputs returns two tensors. 

(This is likely the source of many current test failures.)

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -3235,8 +3287,34 @@ def reference_lgamma(x):
     ]
     op_db = op_db + op_db_scipy_reference
 
+for op in (('add', 'add'), ('mul', 'multiply'), ('div', 'divide')):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3290]: This is clever but let's just add these ops by hand to the op_db list instead of programmatically enumerating them. I imagine that `div` will require special handling, for example.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3290]: How do you want to handle div in this PR?

One option would be to have a domain, but allow specifying a "left domain" or "input domain" and a "right domain" or "right domain." Then divide()'s divisor could be specified to always be greater than 0 + eps, for example. That wouldn't allow generating negative divisors, however. Maybe a follow-up PR could resolve that issue, possibly by updating make_tensor() with a refinement like "bounded away from zero," or even support for more arbitrary domains that include gaps?

**Review by mruberry — COMMENTED** [review]
Hey @mattip! This is a solid start. I've made a few suggestions, particularly simplifying the initial complexity by limiting the new tests.

One remaining challenge for this PR is the sample inputs function. The unary sample inputs func doesn't produce two tensors, so it's causing errors. And while the unary sample inputs function is very simple, the binary sample inputs for these ops are a little more complicated:

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L3569

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L3581

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L3592

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L3606

The binary sample inputs func should subsume these method_tests() and let us delete them.

Also, div already has several OpInfos here:

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L2234

So I think we want to convert those to BinaryUfuncInfos, and update its sample inputs to remove the old method_tests() entries (if possible, maybe that's for a follow-up PR):

https://github.com/pytorch/pytorch/blob/c411017a41988e9c5184279c1ec7dd7ef4e1a6fe/torch/testing/_internal/common_methods_invocations.py#L615

Ping me on Slack if you'd like to discuss this some more.

**Review by mattip — COMMENTED** [review]
I added a `sample_inputs` function, but it is not clear to me how it generates examplar tensors.

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -345,6 +345,94 @@ def __init__(self,
...
+    return (
+        # empty tensors
+        SampleInput((make_tensor((), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 1d - 1d
+        SampleInput((make_tensor((L), device, dtype,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:363]: Let's swap L for M to lower test time slightly (we may revise our input sizes again after @soulitzer improves grad{grad}check performance)

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -345,6 +345,94 @@ def __init__(self,
...
+                                 requires_grad=requires_grad))),
+        # 2d - 1d
+        SampleInput((make_tensor((L, L + 5), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 2d - 2d
+        SampleInput((make_tensor((L, L + 5), device, dtype,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:384]: (M, M) is probably OK here

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -345,6 +345,94 @@ def __init__(self,
...
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L, L + 5), device, dtype,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:387]: If we look at the method_tests() for add:

```
('add', (S, S, S), ((S, S, S),), '', (True,)),
('add', (S, S, S), ((S, S),), 'broadcast_rhs', (True,)),
('add', (S, S), ((S, S, S),), 'broadcast_lhs', (True,)),
('add', (S, 1, S), ((M, S),), 'broadcast_all', (True,)),
('add', (), ((),), 'scalar', (True,)),
('add', (S, S, S), ((),), 'scalar_broadcast_rhs', (True,)),
('add', (), ((S, S, S),), 'scalar_broadcast_lhs', (True,)),
('add', (S, S, S), (3.14,), 'constant', (True,)),
('add', (), (3.14,), 'scalar_constant', (True,)),
('add', (S, S, S), (3.14j,), 'complex_scalar_constant', (True,)),
```

I would expect this to have the same. That is (I've re-ordered the above cases):

```
# scalar tensor x Python scalar (if complex. python scalar should be complex, if int, python scalar should be int)
('add', (), (3.14,), 'scalar_constant', (True,)),

# scalar tensor x scalar tensor
('add', (), ((),), 'scalar', (True,)),

# scalar tensor x 3d
('add', (), ((S, S, S),), 'scalar_broadcast_lhs', (True,)),

# 3d scalar tensor
('add', (S, S, S), ((),), 'scalar_broadcast_rhs', (True,)),

# 3d x python scalar (python scalar should be typed as per dtype)
('add', (S, S, S), (3.14j,), 'complex_scalar_constant', (True,)),
('add', (S, S, S), (3.14,), 'constant', (True,)),

# 3d x 2d
('add', (S, S, S), ((S, S),), 'broadcast_rhs', (True,)),

# 2d x 3d
('add', (S, S), ((S, S, S),), 'broadcast_lhs', (True,)),

# 3d x 3d
('add', (S, S, S), ((S, S, S),), '', (True,)),

# 3d x 2d broadcasting
('add', (S, 1, S), ((M, S),), 'broadcast_all', (True,)),
```

One reasonably quick way of creating these cases is to do something like

```
test_cases = (
  ((S, 1, S), (M, S)),
  ...
)

for test_case in test_cases:
  # instantiate SampleInput here
```

Using that pattern would require adding the Python scalar test case separately. Note that we don't support the first argument to an operation not being a tensor currently, so this should probably limit itself to tensor x python scalar and not pursue python scalar x tensor for now.

Does that make sense?

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -3235,8 +3323,58 @@ def reference_lgamma(x):
...
+                     # skip subtraction of bool types
+                     dtypes=all_types_and_complex(),
+                     dtypesIfCUDA=all_types_and_complex_and(torch.half),
+                     test_inplace_grad=True,
+                     supports_out=True,
+                     sample_inputs_func=sample_inputs_binary,
+                     decorators=(precisionOverride({torch.bfloat16: 5e-1,
+                                                    torch.float16: 5e-1}),),
+                     ),
+    _BinaryUfuncInfo('div',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3361]: I think this PR wants to modify the existing div OpInfos to be _BinaryUfuncInfos, and not create a new one?

**Comment by mruberry** [comment]
> I added a `sample_inputs` function, but it is not clear to me how it generates examplar tensors.

I tried to elaborate. Let me know if that's clear. If not maybe we can Slack or VC?

File: torch/testing/_internal/common_methods_invocations.py (hunk 12 above)
Comment (mattip) [reply torch/testing/_internal/common_methods_invocations.py:3361]: yes, will do

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (mattip) [reply torch/testing/_internal/common_methods_invocations.py:363]: ok

File: torch/testing/_internal/common_methods_invocations.py (hunk 10 above)
Comment (mattip) [reply torch/testing/_internal/common_methods_invocations.py:384]: ok

File: torch/testing/_internal/common_methods_invocations.py (hunk 11 above)
Comment (mattip) [reply torch/testing/_internal/common_methods_invocations.py:387]: yes

Comment (mattip) [reply torch/testing/_internal/common_methods_invocations.py]: Used the paradigm in 4ae138e, without python scalars yet

**Comment by mattip** [comment]
Closing, abandoned.
